Kekbranch

Scrndles opened the pull request:
This pull request contains a partially implemented headshot damage toggle checkbox for the info panel at the trader.
poosh could you look at the ScrnGUIBuyWeaponInfoPanel in this branch (Around line 190), currently the implementation used to get the headshot multiplier out of the different weapons projectile type doesn't seem to be a good one and is broken for ZED gun projectiles, flare projectiles and other weapon types it doesn't know about yet. Perhaps a more robust solution exists that can find impact damage type and headshot damage multiplier for any projectile class but I don't know how to write something like that.
Other than that the branch has some more complete features, here is the full list.
Added new flare effect classes, tweaked flare cloud effect, added head seeking to seeker six SE
Added 3D scope sensitivity adjust to ScrN Settings
Added light brightness and radius decay effect to flare explosion effect.
Implemented partially working headshot damage calculation for weapon buy info panel
Added a faked gas can to chainsaw during reload that also disappears when exploded

poosh commented on Classes/ScrnGUIBuyWeaponInfoPanel.uc:
```diff
@@ -168,11 +175,47 @@ function LoadStats(GUIBuyable NewBuyable, byte FireMode, optional bool bSetTopVa
         BaseDmg = ProjClass.default.Damage;
         DamType = ProjClass.default.MyDamageType;
+        KFDamType = class<KFWeaponDamageType>(DamType); //usually they are roballisticprojectile
+        DamType = ProjClass.default.MyDamageType;
```
Duplicate line

poosh commented on Classes/ScrnGUIBuyWeaponInfoPanel.uc:
```diff
@@ -168,11 +175,47 @@ function LoadStats(GUIBuyable NewBuyable, byte FireMode, optional bool bSetTopVa
+            HSMult = class<M99Bullet>(ProjClass).default.HeadShotDamageMult;
+        }
+        //handle headshot damage calculation for Crossbow and derived projectiles
+        if ( class<CrossbowArrow>(ProjClass) != none)
```
add `else` here and below. Because projectile can't be M99, Crossbow, and M79 at the same time.

Scrndles replied:
Should I still be using a method like this to get headshot damage for different projectile types, it's like hardcoded to recognize every different projectile type when all it needs to do is to check if the projectile has a ImpactDamage and ImpactDamageType. I would have written something like that but I don't know how to.

poosh commented on Classes/ScrnGUIBuyWeaponInfoPanel.uc:
```diff
@@ -204,6 +249,13 @@ function LoadStats(GUIBuyable NewBuyable, byte FireMode, optional bool bSetTopVa
+        HSMult *= Perk.static.GetHeadShotDamMulti(KFPRI, KFPawn(PlayerOwner().Pawn), DamType);
+
+    if (bHSDamage)
+    PerkedValue = PerkedValue * HSMult;
```
missing tab

poosh commented on Classes/ScrnGUIBuyWeaponInfoPanel.uc:
```diff
@@ -212,7 +264,7 @@ function LoadStats(GUIBuyable NewBuyable, byte FireMode, optional bool bSetTopVa
                 PerkedValue += Perk.static.AddDamage(KFPRI, none, KFPawn(PlayerOwner().Pawn), dmg, class<LAWProj>(ProjClass).default.ImpactDamageType); 
             else 
-                PerkedValue += dmg; 
+                PerkedValue += dmg;            
```
remove trailing spaces

poosh commented on Classes/ScrnS6SeekingRocket.uc:
```diff
@@ -1,5 +1,42 @@
+    {
+        //if target has a headbone use those coords, otherwise lock onto position
+        if (Pawn(Seeking).HeadBone != '')
+        LocalSeekingLocation = Seeking.GetBoneCoords(Pawn(Seeking).HeadBone).origin; //get HeadBone name and location
```
A dedicated server is not playing most of the animations, hence the head location on server and client will differ (that's the main reason why headshots are so buggy in KF1). This may lead to situations where projectiles are flying to the different coordinates on server and client. Especially this is bad when the projectile hits something on the way on the server side but flies through on the client.

I agree that this effect is cool, but it cannot be used in KF1 unless we introduce client-side hit detection.

Scrndles replied:
I added this only so that locked on projectiles will get double explosive damage from hitting upper chest shots, actually hitting the head wasn't intended. An initial test version seeking the Pawn.Location with Z+200 could work for large zeds but it would miss crawlers, and it wouldn't match zeds of different height as well as actually using the headbone location. So as long as using headbone coords is good enough to hit high enough to score double damage I don't mind. I think it can be added unless the headbone coords can be so far off that it makes locked rockets miss the zed entirely.

poosh commented on Classes/ScrnPlayerController.uc:
```diff
@@ -23,7 +23,8 @@ var byte VeterancyChangeWave; // wave number, when player has changed his perk
 var globalconfig bool bAlwaysDisplayAchProgression; // always display a message on any achievement progress
-var globalconfig int AchGroupIndex;
+var globalconfig int AchGroupIndex; 
+var globalconfig int Custom3DScopeSens;
```
Nice feature!

poosh commented on Classes/ScrnM4203AssaultRifle.uc:
```diff
@@ -8,8 +8,44 @@ var         name             ReloadShortAnim;
+{
+	local int m;
+    local vector DrawOffset;
+	if (Instigator == None)
```
Tabs instead of spaces... again :)

Scrndles replied:
I fixed it, seems its caused because I used webui to upload or edit this. Guess I can't use it anymore.

poosh commented on Classes/ScrnFlareRevolverImpact.uc:
```diff
@@ -0,0 +1,259 @@
+{
+    LightRadius = LightRadius*0.8;
+    LightBrightness = LightBrightness*0.8;
+    if (LightRadius < 2)
```
Also, disable the timer after turning off the light.

poosh commented on Classes/ScrnChainsaw.uc:
```diff
@@ -164,7 +181,8 @@ simulated function bool ConsumeAmmo( int Mode, float Load, optional bool bAmount
     super.ClientReload();
-
+    if (FakedGasCan != none)
+    FakedGasCan.SetDrawScale(2); //test
```
Missing indent here and below

poosh commented on Classes/ScrnChainsaw.uc:
```diff
@@ -177,7 +195,8 @@ simulated function ClientReload()
     super.ActuallyFinishReloading();
-
+    if (FakedGasCan != none)
+    FakedGasCan.SetDrawScale(0.01); //test
```
Attached actors with modified relative location or rotation do significant performance impact while drawing. I'm not sure if that has any effect on modern computers, but for toaster-players set `FakedGasCan.bHidden` to avoid drawing and potential framerate drop.

Scrndles replied:
I don't know how much performance impact it has, but I tested the Neostead 2000 (It's in LaS Weapon Pack) in solo with uncapped frames. The Neostead 2000 has 8 attached fakedshells each with individually specified location and rotation. I get 400fps looking in a corner, but when I switch to M4 M203 it changes to 380fps. It's not a very indicative test but I think the performance impact still isn't significant.
I can add a var for toaster players just in case though, but where should it be put?

poosh replied:
On second thought, a first-person weapon model is supposed to be high-detail, so there shouldn't be much performance impact even on toasters. Leave the code as is.

Scrndles commented:
Updated list of changes
Attempted fix for AA12 tactical reload faked shell display for golden skins
Added faked gas can to Chainsaw reload to highlight gas explosion mechanic
Added tactical reload to 9mm, Handcannon and MK23
Visual improvements to handcannon's slide, and 9mm and handcannon's hammer
Bullet in old mag for FN FAL tactical reload now visible
New flare explosion effect
New flare cloud effect
Added headshot damage display checkbox to weapon buy info panel
Fixed M4 M203's sight fix to work at all angles
Updated M4 and M4 M203 fire animation rate to match actual firerates
Added tactical reload to MAC10, Thompson Drum SMG and MP5M Medic Gun
Added 3d scope sensitivity adjust slider in ScrN User Settings tab (Adjustable from 1 to 75, default KF value is 24)
Tweaked Seeker Six SE rocket to seek target's head instead of body
SCAR MK17's charging handle locks back on last round fired
Winchester and Trenchgun firing animation doesn't show another round getting chambered when firing last round
